Invasive testing strategy

gilles-peskine-arm opened the pull request:
Documentation only. Explore potential strategies for invasive testing. Formulate some rules and guideline. Succinctly describe some solutions.

This is a proposal submitted for discussion.

I deliberately included solutions with “TODO” in them for things that we should do, but aren't doing yet.

I've changed headers to define macros mentioned explicitly in the strategy document so that one can start implementing parts of the strategy based on this PR. But implementing the proposed strategy is out of scope of this PR.

gilles-peskine-arm commented on docs/architecture/testing/mbed-crypto-invasive-testing.md:
```diff
@@ -0,0 +1,352 @@
+
+Solution: TODO.
+
+See the [secure element driver interface test strategy](driver-interface-test-strategy.html) for more information.
```
Note: currently in https://github.com/ARMmbed/mbed-crypto/pull/246. [Rendered HTML](https://github.com/gilles-peskine-arm/mbed-crypto/blob/psa-se_driver-interface_test_strategy/docs/architecture/testing/driver-interface-test-strategy.md).

mpg commented:
I read the document and it looks very good.

I'd just like to check if I understood something right on a concrete example. Sometimes we have internal functions that we'd like to unit-test because that's the easiest and most efficient way. For example, `mbedtls_ssl_dtls_replay_update()` and `mbedtls_ssl_dlts_replay_check()`. Currently they're unconditonaly non-static (visible by the linker) and declared in `ssl_internal.h`, which is included from the file defining the test functions, but is not supposed to be included by applications.

That's mostly satisfying except that making the functions static would allow more compiler optimizations and possibly reduce the code size. This is a serious concern, as in another instance in the baremetal branch we observed significant reductions in code size (several hundred bytes) just by making a bunch of functions static (see https://github.com/ARMmbed/mbedtls-restricted/pull/652 for details).

IIUC, the proposed rules would encourage us to:

* in non- `MBEDTLS_TEST_HOOKS` builds: make the functions static (allowing the compiler to optimize as fully as it can), but then skip the associated unit tests in the test suite.
* in `MBEDTLS_TEST_HOOKS`-enabled builds: make the functions non-static (potentially increasing the code size) and build and run the associated unit tests as we do now.

Is that correct?

gilles-peskine-arm commented:
@mpg Thanks for the feedback. This is a use case that I hadn't thought of. I agree with your solution. I'll add a paragraph about this to the document.

If a function isn't part of the public API, I encourage declaring it in a header in `library`, not in a public header. We do that for PSA (see `library/psa_*.h`); most of these are exposed for cross-module references, but `psa_crypto_invasive.h` is intended for test only and would be wrapped in a `MBEDTLS_TEST_HOOKS` guard with the new rules.

mpg commented:
> If a function isn't part of the public API, I encourage declaring it in a header in library, not in a public header.

dgreen-arm commented on docs/architecture/testing/mbed-crypto-invasive-testing.md:
```diff
@@ -0,0 +1,354 @@
+* Portability: tests should work on every platform. Skipping tests on certain platforms may hide errors that are only apparent on such platforms.
+* Maintenability: tests should only enforce the documented behavior of the product, to avoid extra work when the product's internal or implementation-specific behavior changes. We should also not give the impression that whatever the tests check is guaranteed behavior of the product which cannot change in future versions.
+
+Where those goals conflit, we should at least mitigate the goals that cannot be fulfilled, and document the architectural choices and their rationale.
```
Typo: `conflit` -> `conflict`

dgreen-arm commented on docs/architecture/testing/mbed-crypto-invasive-testing.md:
```diff
@@ -0,0 +1,354 @@
+* Correctness: we want to test the actual product, not a modified version, since conclusions drawn from a test of a modified product may not apply to the real product.
+* Effacement: the product should not include features that are solely present for test purposes, since these increase the attack surface and the code size.
+* Portability: tests should work on every platform. Skipping tests on certain platforms may hide errors that are only apparent on such platforms.
+* Maintenability: tests should only enforce the documented behavior of the product, to avoid extra work when the product's internal or implementation-specific behavior changes. We should also not give the impression that whatever the tests check is guaranteed behavior of the product which cannot change in future versions.
```
Typo: `Maintenability` -> `Maintainability`

Also appears in a few other places.

gilles-peskine-arm commented:
I rebased on top of `development` to resolve the conflict in `docs/architecture/Makefile` and also fixed the typos that Darryl pointed out. No semantic change.

gilles-peskine-arm commented:
Moved to https://github.com/ARMmbed/mbedtls/pull/3121
